### PR TITLE
check sequence id before updating resource status.

### PR DESCRIPTION
**Pull request by morvencao**
This PR aims to verify the sequence ID (generated using https://github.com/bwmarrin/snowflake) before updating the resource status. This is done to maintain the order of resource status updates.

ref: https://issues.redhat.com/browse/OCM-5359

/assign @qiujian16 @skeeey @clyang82

### CODE REVIEW

**Comment by morvencao** [comment]
Appended test results:

```bash
# make test && make test-integration
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -v  \
	./pkg/... \
	./cmd/...
EMPTY pkg/api
EMPTY pkg/api/openapi
EMPTY pkg/api/presenters
EMPTY pkg/auth
EMPTY pkg/client/cloudevents
EMPTY pkg/client/ocm
PASS pkg/config.TestConfigReadStringFile (0.00s)
PASS pkg/config.TestConfigReadIntFile (0.00s)
PASS pkg/config.TestConfigReadBoolFile (0.00s)
PASS pkg/config.TestConfigReadQuotedFile (0.00s)
PASS pkg/config
PASS pkg/controllers.TestControllerFramework (0.00s)
PASS pkg/controllers
EMPTY pkg/dao
EMPTY pkg/dao/mocks
EMPTY pkg/db
EMPTY pkg/db/db_context
EMPTY pkg/db/db_session
EMPTY pkg/db/migrations
EMPTY pkg/db/mocks
EMPTY pkg/db/transaction
PASS pkg/errors.TestErrorFormatting (0.00s)
PASS pkg/errors.TestErrorFind (0.00s)
PASS pkg/errors
EMPTY pkg/handlers
EMPTY pkg/logger
PASS pkg/services.TestSQLTranslation (0.01s)
PASS pkg/services.TestResourceFindByConsumerID (0.00s)
PASS pkg/services.TestValidateNewManifest/validated (0.00s)
PASS pkg/services.TestValidateNewManifest/no_apiVersion (0.00s)
PASS pkg/services.TestValidateNewManifest/no_kind (0.00s)
PASS pkg/services.TestValidateNewManifest/no_name (0.00s)
PASS pkg/services.TestValidateNewManifest/has_generate_name (0.00s)
PASS pkg/services.TestValidateNewManifest/has_resource_version (0.00s)
PASS pkg/services.TestValidateNewManifest/has_deletion_grace_period_seconds (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_namespace (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version_(no_version) (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version_(no_version)#01 (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version (0.00s)
PASS pkg/services.TestValidateNewManifest (0.00s)
PASS pkg/services.TestValidateUpdateManifest/validated (0.00s)
PASS pkg/services.TestValidateUpdateManifest/apiVersion_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/kind_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/name_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/namespace_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest (0.00s)
PASS pkg/services
EMPTY pkg/util
EMPTY cmd/maestro
EMPTY cmd/maestro/agent
PASS cmd/maestro/environments.TestLoadServices (0.27s)
PASS cmd/maestro/environments
EMPTY cmd/maestro/migrate
EMPTY cmd/maestro/servecmd
EMPTY cmd/maestro/server
EMPTY cmd/maestro/server/logging

DONE 28 tests in 8.480s
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -ldflags -s -v -timeout 1h  \
		./test/integration
I0110 03:31:04.554279   82033 framework.go:76] Initializing testing environment
I0110 03:31:04.752039   82033 framework.go:161] Using Mock OCM Authz Client
I0110 03:31:04.753260   82033 framework.go:199] Disabling Sentry error reporting
{"level":"info","ts":1704857464.7535057,"logger":"fallback","caller":"v2@v2.0.0-20231030012137-0836a524e995/protocol.go:124","msg":"subscribing to topics: map[sources/clusters/+/specresync:{1 0 false false} sources/maestro-integration/clusters/+/status:{1 0 false false}]"}
I0110 03:31:04.760144   82033 api_server.go:145] Serving without TLS at localhost:8000
PASS test/integration.TestConsumerGet (0.22s)
PASS test/integration.TestConsumerPost (0.20s)
PASS test/integration.TestConsumerPaging (0.19s)
PASS test/integration.TestControllerRacing (1.46s)
PASS test/integration.TestControllerReconcile (1.28s)
PASS test/integration.TestControllerSync (1.27s)
PASS test/integration.TestResourceGet (0.20s)
PASS test/integration.TestResourcePost (1.17s)
PASS test/integration.TestResourcePatch (0.25s)
PASS test/integration.TestResourcePaging (0.34s)
PASS test/integration.TestResourceListSearch (0.27s)
PASS test/integration.TestUpdateResourceWithRacingRequests (0.24s)
I0110 03:31:11.867584   82033 api_server.go:151] Web server terminated
PASS test/integration (cached)

DONE 12 tests in 0.904s
```

**Comment by qiujian16** [comment]
LGTM

File: pkg/services/resource.go [hunk 1]
```diff
@@ -154,6 +155,41 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
+		return nil, errors.GeneralError("Unable to marshal resource status: %s", err)
+	}
+	foundStatus := &api.ResourceStatus{}
+	if err := json.Unmarshal(foundStatusJSON, foundStatus); err != nil {
+		return nil, errors.GeneralError("Unable to unmarshal resource status: %s", err)
+	}
+
+	sequenceID, foundSequenceID := "", ""
+	if resourceStatus.ReconcileStatus == nil && foundStatus.ReconcileStatus != nil {
+		sequenceID = resourceStatus.ReconcileStatus.SequenceID
```
Comment (skeeey) [inline pkg/services/resource.go:178]: nil point?

File: pkg/services/resource.go [hunk 2]
```diff
@@ -154,6 +155,41 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
+	}
+
+	foundStatusJSON, err := json.Marshal(found.Status)
+	if err != nil {
+		return nil, errors.GeneralError("Unable to marshal resource status: %s", err)
+	}
+	foundStatus := &api.ResourceStatus{}
+	if err := json.Unmarshal(foundStatusJSON, foundStatus); err != nil {
+		return nil, errors.GeneralError("Unable to unmarshal resource status: %s", err)
+	}
```
Comment (skeeey) [inline pkg/services/resource.go:174]: nit: cloud we add a help function to convert the status object?

File: pkg/services/resource.go [hunk 3]
```diff
@@ -154,6 +155,41 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
+
+	sequenceID, foundSequenceID := "", ""
+	if resourceStatus.ReconcileStatus == nil && foundStatus.ReconcileStatus != nil {
+		sequenceID = resourceStatus.ReconcileStatus.SequenceID
+		foundSequenceID = foundStatus.ReconcileStatus.SequenceID
+	}
+
+	if sequenceID != "" && foundSequenceID != "" {
```
Comment (skeeey) [inline pkg/services/resource.go:182]: we may not need this, I think the sequenceID should always exist

File: pkg/services/resource.go (hunk 1 above)
Comment (morvencao) [reply pkg/services/resource.go:178]: add check for nil pointer.

File: pkg/services/resource.go (hunk 2 above)
Comment (morvencao) [reply pkg/services/resource.go:174]: add helper function and testings.

File: pkg/services/resource.go (hunk 3 above)
Comment (morvencao) [reply pkg/services/resource.go]: removed this check.

**Comment by skeeey** [comment]
LGTM

**Comment by morvencao** [comment]
Test results:

```bash
# make test && make test-integration
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -v  \
	./pkg/... \
	./cmd/...
PASS pkg/api.TestJSONMapStausToResourceStatus/empty (0.00s)
PASS pkg/api.TestJSONMapStausToResourceStatus/content_status (0.00s)
PASS pkg/api.TestJSONMapStausToResourceStatus/reconcile_status (0.00s)
PASS pkg/api.TestJSONMapStausToResourceStatus/reconcile_status_with_deleted_Condition (0.00s)
PASS pkg/api.TestJSONMapStausToResourceStatus (0.00s)
PASS pkg/api.TestResourceStatusToJSONMap/empty (0.00s)
PASS pkg/api.TestResourceStatusToJSONMap/content_status (0.00s)
PASS pkg/api.TestResourceStatusToJSONMap/reconcile_status (0.00s)
PASS pkg/api.TestResourceStatusToJSONMap/reconcile_status_with_deleted_Condition (0.00s)
PASS pkg/api.TestResourceStatusToJSONMap (0.00s)
PASS pkg/api (cached)
EMPTY pkg/api/openapi
EMPTY pkg/api/presenters
EMPTY pkg/auth
EMPTY pkg/client/cloudevents
EMPTY pkg/client/ocm
PASS pkg/config.TestConfigReadStringFile (0.00s)
PASS pkg/config.TestConfigReadIntFile (0.00s)
PASS pkg/config.TestConfigReadBoolFile (0.00s)
PASS pkg/config.TestConfigReadQuotedFile (0.00s)
PASS pkg/config (cached)
PASS pkg/controllers.TestControllerFramework (0.00s)
PASS pkg/controllers (cached)
EMPTY pkg/dao
EMPTY pkg/dao/mocks
EMPTY pkg/db
EMPTY pkg/db/db_context
EMPTY pkg/db/db_session
EMPTY pkg/db/migrations
EMPTY pkg/db/mocks
EMPTY pkg/db/transaction
PASS pkg/errors.TestErrorFormatting (0.00s)
PASS pkg/errors.TestErrorFind (0.00s)
PASS pkg/errors (cached)
EMPTY pkg/handlers (cached)
EMPTY pkg/logger
PASS pkg/services.TestSQLTranslation (0.02s)
PASS pkg/services.TestResourceFindByConsumerID (0.00s)
PASS pkg/services.TestValidateNewManifest/validated (0.00s)
PASS pkg/services.TestValidateNewManifest/no_apiVersion (0.00s)
PASS pkg/services.TestValidateNewManifest/no_kind (0.00s)
PASS pkg/services.TestValidateNewManifest/no_name (0.00s)
PASS pkg/services.TestValidateNewManifest/has_generate_name (0.00s)
PASS pkg/services.TestValidateNewManifest/has_resource_version (0.00s)
PASS pkg/services.TestValidateNewManifest/has_deletion_grace_period_seconds (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_namespace (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version_(no_version) (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version_(no_version)#01 (0.00s)
PASS pkg/services.TestValidateNewManifest/wrong_api_version (0.00s)
PASS pkg/services.TestValidateNewManifest (0.00s)
PASS pkg/services.TestValidateUpdateManifest/validated (0.00s)
PASS pkg/services.TestValidateUpdateManifest/apiVersion_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/kind_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/name_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest/namespace_mismatch (0.00s)
PASS pkg/services.TestValidateUpdateManifest (0.00s)
PASS pkg/services (cached)
EMPTY pkg/util
EMPTY cmd/maestro
EMPTY cmd/maestro/agent
PASS cmd/maestro/environments.TestLoadServices (0.38s)
PASS cmd/maestro/environments (cached)
EMPTY cmd/maestro/migrate
EMPTY cmd/maestro/servecmd
EMPTY cmd/maestro/server
EMPTY cmd/maestro/server/logging

DONE 38 tests in 2.559s
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -ldflags -s -v -timeout 1h  \
		./test/integration
I0112 03:10:46.134525  768806 framework.go:76] Initializing testing environment
I0112 03:10:46.346614  768806 framework.go:161] Using Mock OCM Authz Client
I0112 03:10:46.348517  768806 framework.go:199] Disabling Sentry error reporting
{"level":"info","ts":1705029046.3487647,"logger":"fallback","caller":"v2@v2.0.0-20231030012137-0836a524e995/protocol.go:124","msg":"subscribing to topics: map[sources/clusters/+/specresync:{1 0 false false} sources/maestro-integration/clusters/+/status:{1 0 false false}]"}
I0112 03:10:46.358289  768806 api_server.go:145] Serving without TLS at localhost:8000
PASS test/integration.TestConsumerGet (0.28s)
PASS test/integration.TestConsumerPost (0.28s)
PASS test/integration.TestConsumerPaging (0.19s)
PASS test/integration.TestControllerRacing (1.66s)
PASS test/integration.TestControllerReconcile (1.26s)
PASS test/integration.TestControllerSync (1.26s)
PASS test/integration.TestResourceGet (0.17s)
PASS test/integration.TestResourcePost (2.17s)
PASS test/integration.TestResourcePatch (0.22s)
PASS test/integration.TestResourcePaging (0.38s)
PASS test/integration.TestResourceListSearch (0.28s)
PASS test/integration.TestUpdateResourceWithRacingRequests (0.40s)
I0112 03:10:54.945428  768806 api_server.go:151] Web server terminated
PASS test/integration

DONE 12 tests in 12.503s
```

File: pkg/api/resource_types.go [hunk 4]
```diff
@@ -1,8 +1,10 @@
 package api
 
 import (
+	"encoding/json"
 	"strconv"
 
+	"github.com/openshift-online/maestro/pkg/errors"
```
Comment (clyang82) [inline pkg/api/resource_types.go:7]: could you follow the standard for the order of imports? I will introduce `gci` soon to format the orders.

**Review by clyang82 — APPROVED** [review]
LGTM

File: pkg/api/resource_types.go (hunk 4 above)
Comment (morvencao) [reply pkg/api/resource_types.go]: done.

**Pull request merged**